Assets/MixedRealityToolkit/Utilities/MixedRealityServiceRegistry.cs

```diff
@@ -0,0 +1,222 @@
+            {
+                return false;
+            }
+
+            if (!object.ReferenceEquals(serviceInstance, tempService))
+            {
+                return false;
+            }
+
+            return RemoveServiceInternal(typeof(T), serviceInstance, registrar);
+        }
+
+        /// <summary>
+        /// Removes an <see cref="IMixedRealityService"/> instance from the registry.
+        /// </summary>
+        /// <typeparam name="T">The interface type of the service being removed.</typeparam>
+        /// <param name="name">The friendly name of the service to remove.</param>
+        /// <returns>
+        /// True if the service was successfully removed, false otherwise.
+        /// </returns>
+        public static bool RemoveService<T>(string name) where T : IMixedRealityService
+        {
+            T tempService;            
+            IMixedRealityServiceRegistrar registrar;
+
+            if (!TryGetService<T>(out tempService, out registrar, name))
+            {
+                return false;
+            }
+
+            return RemoveServiceInternal(typeof(T), tempService, registrar);
+        }
+
+        /// <summary>
+        /// Removes an <see cref="IMixedRealityService"/> instance from the registry.
+        /// </summary>
+        /// <param name="interfaceType">The interface type of the service being removed.</param>
+        /// <param name="serviceInstance">Instance of the service to remove.</param>
+        /// <param name="registrar">Instance of the registrar manages the service.</param>
+        /// <returns>
+        /// True if the service was successfully removed, false otherwise.
+        /// </returns>
+        private static bool RemoveServiceInternal(
+            Type interfaceType,
+            IMixedRealityService serviceInstance,
+            IMixedRealityServiceRegistrar registrar)
+        {
+            List<KeyValuePair<IMixedRealityService, IMixedRealityServiceRegistrar>> services = registry[interfaceType];
+
+            return services.Remove(new KeyValuePair<IMixedRealityService, IMixedRealityServiceRegistrar>(serviceInstance, registrar));
+        }
+
+        /// <summary>
+        /// Gets the instance of the requested service from the registry.
+        /// </summary>
+        /// <typeparam name="T">The interface type of the service being requested.</typeparam>
+        /// <param name="serviceInstance">Output parameter to receive the requested service instance.</param>
+        /// <param name="registrar">Output parameter to receive the registrar that loaded the service instance.</param>
+        /// <param name="name">Optional name of the service.</param>
+        /// <returns>
+        /// True if the requested service is being returned, false otherwise.
+        /// </returns>
+        public static bool TryGetService<T>(
+            out T serviceInstance,
+            out IMixedRealityServiceRegistrar registrar,
```

Comment:
It's weird seeing outparams being first, instead of last - in other languages that support this type of mechanic most of the coding guidelines I've seen have been to put the input first, output last.

Other examples of code from warnings/guidelines:
[https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1021-avoid-out-parameters?view=vs-2019](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1021-avoid-out-parameters?view=vs-2019)
[https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/out-parameter-modifier](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/out-parameter-modifier)

Seem to list it last as well.

Reply:
I guess in this case it's interesting because name is optional.

---
In reply to: [279109153](https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/4099#discussion_r279109153) [](ancestors = 279109153)

Reply:
we discussed internally and decided to leave as is